### PR TITLE
MGMTBUGSM-412: validate pull secret in kube-api cluster deployment

### PR DESCRIPTION
**Pull request by slaviered**
## List all the issues related to this PR

- [ ] New Feature 
- [ ] Enhancement 
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [ ] Cloud
- [ ] Operator Managed Deployments
- [x] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [x] No tests needed

## Assignees

/cc @
/cc @

## Checklist

- [ ] Title and description added to both, commit and PR.
- [ ] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [ ] Reviewers have been listed
- [ ] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: internal/controller/controllers/clusterdeployments_controller.go [hunk 1]
```diff
@@ -230,7 +231,7 @@ func (r *ClusterDeploymentsReconciler) Reconcile(origCtx context.Context, req ct
 
 	if swag.StringValue(cluster.Kind) == models.ClusterKindCluster && !clusterInstall.Spec.HoldInstallation {
 		// Day 1
-		pullSecret, err := getAndLabelPullSecret(ctx, r.Client, r.APIReader, clusterDeployment.Spec.PullSecretRef, req.Namespace)
+		pullSecret, err := r.PullSecretClient.GetValidPullSecret(ctx, getPullSecretKey(req.Namespace, clusterDeployment.Spec.PullSecretRef))
 		if err != nil {
 			log.WithError(err).Error("failed to get pull secret")
```
Comment (danielerez) [inline internal/controller/controllers/clusterdeployments_controller.go:236]: It can also be an invalid secret if we get here. Can we differentiate between a failure the fetch the secret and a case of an invalid secret?  So we could treat an invalid secret as we do for invalid release image: https://github.com/openshift/assisted-service/blob/a5cafb300700e5440504eae1de1cdb4ee0884619/internal/controller/controllers/clusterdeployments_controller.go#L387

File: internal/controller/controllers/clusterdeployments_controller_test.go [hunk 2]
```diff
@@ -3507,12 +3562,13 @@ var _ = Describe("unbindAgents", func() {
...
-			Installer:  mockInstallerInternal,
-			ClusterApi: mockClusterApi,
+			Client:           c,
+			APIReader:        c,
+			Scheme:           scheme.Scheme,
+			Log:              common.GetTestLog(),
+			Installer:        mockInstallerInternal,
+			ClusterApi:       mockClusterApi,
+			PullSecretClient: NewPullSecretHandler(c, c, mockInstallerInternal),
 		}
```
Comment (danielerez) [inline internal/controller/controllers/clusterdeployments_controller_test.go:3573]: Can we add a test for the invalid secret flow.

File: internal/controller/controllers/clusterdeployments_controller.go [hunk 3]
```diff
@@ -230,7 +231,7 @@ func (r *ClusterDeploymentsReconciler) Reconcile(origCtx context.Context, req ct
 
 	if swag.StringValue(cluster.Kind) == models.ClusterKindCluster && !clusterInstall.Spec.HoldInstallation {
 		// Day 1
-		pullSecret, err := getAndLabelPullSecret(ctx, r.Client, r.APIReader, clusterDeployment.Spec.PullSecretRef, req.Namespace)
+		pullSecret, err := r.PullSecretClient.GetValidPullSecret(ctx, getPullSecretKey(req.Namespace, clusterDeployment.Spec.PullSecretRef))
```
[the remark this answers is not in the document]
Comment (slaviered) [reply internal/controller/controllers/clusterdeployments_controller.go:234]: It was added because it needed a state (k8s client and api reader). The creation of the state cleaned the API and allowed it to be re-used from both kube layer and REST

File: internal/controller/controllers/clusterdeployments_controller.go (hunk 1 above)
Comment (slaviered) [reply internal/controller/controllers/clusterdeployments_controller.go:236]: The result is the same: the secret can not be use and the user should check its configurations. I did not change the API dramatically in this PR because it suppose to be a small change

File: internal/controller/controllers/clusterdeployments_controller_test.go (hunk 2 above)
Comment (slaviered) [reply internal/controller/controllers/clusterdeployments_controller_test.go:3573]: I will

File: internal/controller/controllers/clusterdeployments_controller.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (danielerez) [reply internal/controller/controllers/clusterdeployments_controller.go:234]: ok so worth moving later also getSecret func to handler: https://github.com/openshift/assisted-service/blob/cca573243ca975d0bd4d6c4ec212586804bee833/internal/controller/controllers/common.go#L58

File: internal/controller/controllers/clusterdeployments_controller.go (hunk 1 above)
Comment (danielerez) [reply internal/controller/controllers/clusterdeployments_controller.go:236]: Then maybe rephrase the log message to cover both cases, e.g. "missing or invalid Pull Secret"

File: internal/controller/controllers/clusterdeployments_controller.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (slaviered) [reply internal/controller/controllers/clusterdeployments_controller.go:231]: It's an internal utility not and API

File: internal/controller/controllers/clusterdeployments_controller.go (hunk 1 above)
Comment (slaviered) [reply internal/controller/controllers/clusterdeployments_controller.go:233]: After investigating the code there is no need to change the log. There are ample logs in ValidatePullSecret function that describe specifically what went wrong. Then, in addition, there is this logs which is a legacy log. Together, they provide an accurate picture of what is wrong and how to fix it.
